cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql

```diff
@@ -93,11 +93,12 @@ predicate isInvalidPointerDerefSink2(DataFlow::Node sink, Instruction i, string
   )
 }
 
-predicate isConstantSizeOverflowSource(Field f, PointerAddInstruction pai, int delta) {
-  exists(int size, int bound, DataFlow::Node source, DataFlow::InstructionNode sink |
-    FieldAddressToPointerArithmeticFlow::flow(source, sink) and
-    isFieldAddressSource(f, source) and
-    pai.getLeft() = sink.asInstruction() and
+predicate isConstantSizeOverflowSource(Field f, FieldAddressToPointerArithmeticFlow::PathNode fieldSource, PointerAddInstruction pai, int delta) {
+  exists(int size, int bound, FieldAddressToPointerArithmeticFlow::PathNode sink |
+    FieldAddressToPointerArithmeticFlow::flowPath(fieldSource, sink) and
+    isFieldAddressSource(f, fieldSource.getNode()) and
+    pai.getLeft() = sink.getNode().(DataFlow::InstructionNode).asInstruction() and
+    pai.getElementSize() = f.getUnspecifiedType().(ArrayType).getBaseType().getSize() and
```

Comment:
I think we should push the `f.getUnspecifiedType() instanceof ArrayType` conjunct into the `FieldAddressToPointerArithmeticConfig::isSource` predicate (or even better: into the `isFieldAddressSource` predicate) to get a smaller `isSource` predicate for the first dataflow traversal.

Thinking more about this: If we push the restriction saying that `f.getUnspecifiedType()` must be an `ArrayType` into the `isSource` predicate, we can use a flow state to remember the size of the array. Does that not mean that we would be able to handle cases like the one you added:
```cpp
char *charBuf = (char*) arr->buf;
charBuf[MAX_SIZE_BYTES] = 0;
```
then?